platform/linux-generic/odp_packet_io.c

```diff
@@ -1699,33 +1699,31 @@ int odp_pktin_recv_tmo(odp_pktin_queue_t queue, odp_packet_t packets[], int num,
 	while (1) {
 		ret = entry->s.ops->recv(entry, queue.index, packets, num);
 
-		if (ret != 0)
+		if (ret != 0 || wait == 0)
 			return ret;
 
-		if (wait == 0)
-			return 0;
-
-		if (wait != ODP_PKTIN_WAIT) {
-			/* Avoid unnecessary system calls. Record the start time
-			 * only when needed and after the first call to recv. */
-			if (odp_unlikely(!started)) {
-				odp_time_t t;
-
-				t = odp_time_local_from_ns(wait * 1000);
-				started = 1;
-				t1 = odp_time_sum(odp_time_local(), t);
-			}
+		/* Avoid unnecessary system calls. Record the start time
+		 * only when needed and after the first call to recv. */
+		if (odp_unlikely(!started)) {
+			odp_time_t t;
+
+			/* Avoid overflow issues for large wait times */
+			if (wait > odp_pktin_wait_time(UINT64_MAX / 2))
+				wait /= 2;
```

Comment:
Idea of wait_time() is that application calls that once in init time and not per packet, so that expensive division operation can be avoided per packet. So, implementation should not do division (or call wait_time()) per packet either. Use a constant instead:
    if (wait > MAX_WAIT_TIME)
       wait = MAX_WAIT_TIME;